internal/server/snapshot.go

```diff
@@ -0,0 +1,318 @@
+package server
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/tidwall/gjson"
+	"github.com/tidwall/resp"
+	"github.com/tidwall/tile38/internal/collection"
+	"github.com/tidwall/tile38/internal/log"
+)
+
+var errSnapshotLoadFailed = errors.New("snapshot load failed")
+
+const (
+	Id     = "id"
+	Offset = "offset"
+)
+
+// Record of the last snapshot for this dataset
+type SnapshotMeta struct {
+	path string
+
+	mu sync.RWMutex
+
+	_idstr	string
+	_offset	int64
+}
+
+func loadSnapshotMeta(path string) (sm *SnapshotMeta, err error) {
+	sm = &SnapshotMeta{path: path}
+	var jsonStr string
+	var data []byte
+	data, err = ioutil.ReadFile(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return sm, nil
+		} else {
+			return nil, err
+		}
+	}
+
+	jsonStr = string(data)
+	sm._idstr = gjson.Get(jsonStr, Id).String()
+	sm._offset = gjson.Get(jsonStr, Offset).Int()
+
+	return sm, nil
+}
+
+func (sm *SnapshotMeta) save() error {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	m := make(map[string]interface{})
+	if sm._idstr != "" {
+		m[Id] = sm._idstr
+	}
+	if sm._offset != 0 {
+		m[Offset] = sm._offset
+	}
+	data, err := json.MarshalIndent(m, "","\t")
+	if err != nil {
+		return err
+	}
+	data = append(data, '\n')
+	err = ioutil.WriteFile(sm.path, data, 0600)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func connLastSnapshotMeta(conn *RESPConn) (snapshotMeta *SnapshotMeta, err error) {
+	v, e := conn.Do("snapshot latest meta")
+	if e != nil {
+		err = e
+		return
+	}
+	if v.Error() != nil {
+		err = v.Error()
+		return
+	}
+	vals := v.Array()
+	snapshotMeta = &SnapshotMeta{
+		_idstr: vals[0].String(),
+		_offset: int64(vals[1].Integer()),
+	}
+	return
+}
+
+func (s *Server) cmdSnapshotLastMeta(msg *Message) (res resp.Value, err error) {
+	start := time.Now()
+	switch msg.OutputType {
+	case JSON:
+		res = resp.StringValue(
+			fmt.Sprintf(
+				`{"ok":true,"id":"%s","offset":%d,elapsed":"%s"}`,
+				s.snapshotMeta._idstr,
+				s.snapshotMeta._offset,
+				time.Now().Sub(start)))
+	case RESP:
+		res = resp.ArrayValue([]resp.Value{
+			resp.SimpleStringValue(s.snapshotMeta._idstr),
+			resp.IntegerValue(int(s.snapshotMeta._offset)),
+		})
+	}
+	return res, nil
+}
+
+func (s *Server) getSnapshotDir(snapshotIdStr string) string {
+	return filepath.Join(s.dir, "snapshots", snapshotIdStr)
+}
+
+func (s *Server) cmdSaveSnapshot(msg *Message) (res resp.Value, err error) {
+	start := time.Now()
+	snapshotId := rand.Uint64()
+	snapshotIdStr := strconv.FormatUint(snapshotId, 16)
+	if err := s.writeAOF([]string{"SAVESNAPSHOT", snapshotIdStr}, nil); err != nil {
+		log.Errorf("Failed to write AOF for snapshot: %v", err)
+		return NOMessage, errInvalidAOF
+	}
+	go s.doSaveSnapshot(snapshotId, snapshotIdStr, s.aofsz)
+	switch msg.OutputType {
```

Comment:
Would it be worth waiting for the snapshot to complete here so client can know it's done?